ta/remoteproc/remoteproc_core.c

```diff
@@ -0,0 +1,1212 @@
+
+	if (size) {
+		res = TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+					  PTA_RPROC_LOAD_SEGMENT_SHA256,
+					  param_types, params, NULL);
+		if (res != TEE_SUCCESS) {
+			EMSG("Fails to load segment, res = 0x%x", res);
+			return res;
+		}
+	}
+
+	/* Fill the rest of the memory with 0 */
+	if (size < mem_size) {
+		param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					      TEE_PARAM_TYPE_VALUE_INPUT,
+					      TEE_PARAM_TYPE_VALUE_INPUT,
+					      TEE_PARAM_TYPE_VALUE_INPUT);
+		params[1].value.a = da + size;
+		params[2].value.a = mem_size - size;
+		params[3].value.a = 0;
+
+		res = TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+					  PTA_RPROC_SET_MEMORY,
+					  param_types, params, NULL);
+		if (res != TEE_SUCCESS)
+			EMSG("Fails to clear segment, res = 0x%x", res);
+	}
+
+	return res;
+}
+
+static TEE_Result remoteproc_load_elf(struct remoteproc_context *ctx)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t num_img = 0, i = 0;
+	uint8_t img_type = REMOTEPROC_INVALID_TYPE;
+	uint32_t img_size = 0;
+	uint8_t *tlv = NULL;
+	int32_t offset = 0;
+	size_t length = 0;
+	paddr_t rsc_pa = 0;
+	size_t rsc_size = 0;
+
+	res = e32_parse_ehdr(ctx->fw_img, ctx->fw_img_sz);
+	if (res) {
+		EMSG("Failed to parse firmware, res = %#"PRIx32, res);
+		return res;
+	}
+
+	res = get_hash_table(ctx);
+	if (res)
+		return res;
+
+	/* Get the number of firmwares to load */
+
+	res = remoteproc_get_tlv(ctx->tlr, RPROC_TLV_NUM_IMG,
+				 &tlv, &length);
+	if (res)
+		goto out;
+	if (length != sizeof(uint8_t)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
+	}
+
+	num_img = *tlv;
+	ctx->rsc_pa = 0;
+	ctx->rsc_size = 0;
+
+	for (i = 0; i < num_img; i++) {
+		res = get_tlv_images_type(ctx, num_img, i, &img_type);
+		if (res)
+			goto out;
+		if (img_type != REMOTEPROC_ELF_TYPE) {
+			res = TEE_ERROR_BAD_FORMAT;
+			goto out;
+		}
+
+		res = get_tlv_images_size(ctx, num_img, i, &img_size);
+		if (res)
+			goto out;
+
+		res = e32_parser_load_elf_image(ctx->fw_img + offset, img_size,
+						remoteproc_load_segment, ctx);
+		if (res)
+			goto out;
+
+		/* Take opportunity to get the resource table address */
+		res = remoteproc_parse_rsc_table(ctx, ctx->fw_img + offset,
+						 img_size, &rsc_pa, &rsc_size);
+		if (res != TEE_SUCCESS && res != TEE_ERROR_NO_DATA)
+			goto out;
+
+		if (res != TEE_ERROR_NO_DATA) {
+			/*
+			 * Only one resource table is supported, check that no
+			 * other one has been declared in a previous loaded
+			 * firmware.
+			 */
+			if (ctx->rsc_pa || ctx->rsc_size) {
+				EMSG("More than one resource table found");
+				res = TEE_ERROR_BAD_FORMAT;
+				goto out;
+			}
+			ctx->rsc_pa = rsc_pa;
+			ctx->rsc_size = rsc_size;
+		}
+
+		offset += img_size;
+	}
+
+	/*
+	 * If res == TEE_ERROR_NO_DATA then no resource table found.
+	 * Force to TEE_SUCCESS as the resource table is optional.
+	 */
+	res = TEE_SUCCESS;
+
+out:
+	/* TODO: clean-up the memories in case of fail */
```

Comment:
You can clear the memory content. Use `memzero_explicit()` if necessary.

Reply:
No  possible here we don't know which memory should be cleaned. the memory should be cleaned by the platform. storing a list of memory segment filled seems not appropriate. 
The question is does we need to clean the memory , I changed the TODO by a comment with a question.
if yes either we do nothing here or we call a new PTA service to clean-up the memory